controller/gke-cluster-config-handler.go

```diff
@@ -0,0 +1,349 @@
+		logrus.Warnf("cluster [%s] never advanced to creating status, will not delete GKE cluster", config.Name)
+		return config, nil
+	}
+
+	logrus.Infof("deleting cluster [%s]", config.Name)
+
+	//TODO: delete cluster
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	svc, err := getServiceClient(ctx, config.Spec.CredentialContent)
+	if err != nil {
+		return config, err
+	}
+
+	logrus.Debugf("Removing cluster %v from project %v, region/zone %v", config.Spec.ClusterName, config.Spec.ProjectID, config.Spec.Region)
+	operation, err := waitClusterRemoveExp(ctx, svc, config)
+
+	if err != nil && !strings.Contains(err.Error(), "notFound") {
+		return config, err
+	} else if err == nil {
+		logrus.Debugf("Cluster %v delete is called. Status Code %v", config.Spec.ClusterName, operation.HTTPStatusCode)
+	} else {
+		logrus.Debugf("Cluster %s doesn't exist", config.Spec.ClusterName)
+	}
+
+	return config, nil
+}
+
+func (h *Handler) create(config *gkev1.GKEClusterConfig) (*gkev1.GKEClusterConfig, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	svc, err := getServiceClient(ctx, config.Spec.CredentialContent)
+	if err != nil {
+		return config, err
+	}
+
+	operation, err := svc.Projects.Locations.Clusters.Create(
+		locationRRN(config.Spec.ProjectID, config.Spec.Region),
+		generateClusterCreateRequest(config)).Context(ctx).Do()
+
+	if err != nil && !strings.Contains(err.Error(), "alreadyExists") {
+		return config, err
+	}
+
+	if err == nil {
+		logrus.Debugf("Cluster %s create is called for project %s and region/zone %s. Status Code %v",
+			config.Spec.ClusterName, config.Spec.ProjectID, config.Spec.Region, operation.HTTPStatusCode)
+	}
+
+	config = config.DeepCopy()
+	config.Status.Phase = gkeConfigCreatingPhase
+	return h.gkeCC.UpdateStatus(config)
+}
+
+func (h *Handler) checkAndUpdate(config *gkev1.GKEClusterConfig) (*gkev1.GKEClusterConfig, error) {
+
+	//TODO: implement -- currently a noop.
+	return config, nil
+}
+
+func (h *Handler) waitForCreationComplete(config *gkev1.GKEClusterConfig) (*gkev1.GKEClusterConfig, error) {
+	lastMsg := ""
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	for {
+		svc, err := getServiceClient(ctx, config.Spec.CredentialContent)
+		if err != nil {
+			return config, err
+		}
+		cluster, err := svc.Projects.Locations.Clusters.Get(clusterRRN(config.Spec.ProjectID, config.Spec.Region, config.Spec.ClusterName)).Context(ctx).Do()
+		if err != nil {
+			return config, err
+		}
+		if cluster.Status == runningStatus {
+			log.Infof(ctx, "Cluster %v is running", config.Spec.ClusterName)
+			config = config.DeepCopy()
+			config.Status.Phase = gkeConfigActivePhase
+			return h.gkeCC.UpdateStatus(config)
+		}
+		if cluster.Status != lastMsg {
+			log.Infof(ctx, "%v cluster %v......", strings.ToLower(cluster.Status), config.Spec.ClusterName)
+			lastMsg = cluster.Status
+		}
+		time.Sleep(time.Second * 5)
```

Comment:
I think it may be better to use renqueue after instead of sleeping in a for loop.

Reply:
As stated we should probably reenqueue, by returning an error or actually calling enqueue, instead. One reason why is waiting for creation to complete can be relatively long. For the instances where we do want to perform this logic, and some one argue it would be fine to do so here, the Backoff functions from k8s.io/apimachinery are good.